// cucumber/lib/formatter/json_formatter.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _lodash = _interopRequireDefault(require("lodash"));

var _2 = _interopRequireDefault(require("./"));

var _helpers = require("./helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class JsonFormatter extends _2.default {
  constructor(options) {
    super(options);
    options.eventBroadcaster.on('test-run-finished', this.onTestRunFinished.bind(this));
  }

  formatExceptionIfNeeded(obj) {
    if (obj.result.exception) {
      obj.result.exception = (0, _helpers.formatError)(obj.result.exception);
    }
  }

  onTestRunFinished() {
    const testCaseAttempts = this.eventDataCollector.getTestCaseAttempts();
    const testCaseAttemptsData = testCaseAttempts.map(testCaseAttempt => {
      const parsed = (0, _helpers.parseTestCaseAttempt)({
        snippetBuilder: this.snippetBuilder,
        testCaseAttempt: testCaseAttempt
      });
      this.formatExceptionIfNeeded(parsed.testCase);
      parsed.testSteps.forEach(s => this.formatExceptionIfNeeded(s));
      return parsed;
    });
    const data = {
      gherkinDocuments: _lodash.default.values(this.eventDataCollector.gherkinDocumentMap),
      pickles: _lodash.default.values(this.eventDataCollector.pickleMap),
      testCaseAttempts: testCaseAttemptsData
    };
    this.log(JSON.stringify(data, null, 2));
  }

}

exports.default = JsonFormatter;
